data/json/ui/bodypart_status.json

```diff
@@ -1,10 +1,35 @@
 [
+  {
+    "id": "bodypart_status_indicator_template",
+    "type": "widget",
+    "style": "phrase",
+    "phrases": [
+      { "id": "bitten", "text": "bitten", "sym": "B", "color": "yellow" },
+      { "id": "infected", "text": "infected", "sym": "I", "color": "pink" },
+      { "id": "broken", "text": "broken", "sym": "%", "color": "magenta" },
+      { "id": "splinted", "text": "splinted", "sym": "=", "color": "light_gray" },
+      { "id": "bandaged", "text": "bandaged", "sym": "+", "color": "white" },
+      { "id": "disinfected", "text": "disinfected", "sym": "$", "color": "light_green" },
+      { "id": "bleeding", "text": "bleeding", "value": 0, "sym": "b", "color": "light_red" },
+      { "id": "bleeding", "text": "bleeding", "value": 11, "sym": "b", "color": "red" },
+      { "id": "bleeding", "text": "bleeding", "value": 21, "sym": "b", "color": "red_red" }
+    ]
+  },
   {
     "id": "bodypart_status_text",
     "//": "Base widget for showing body part status; needs bodypart field defined in derived widget.",
     "type": "widget",
     "style": "text",
-    "var": "bp_status_text"
+    "var": "bp_status_text",
+    "copy-from": "bodypart_status_indicator_template"
+  },
+  {
+    "id": "bodypart_status_sym_text",
+    "//": "Base widget for body part status symbols; needs bodypart field defined in derived widget.",
+    "type": "widget",
+    "style": "text",
```

Comment:
Although the base style is "phrase", the copy has style "text", and it still works OK right?

If so, there may be no need for a separate "phrase" style; we should just use "text" for all of them.

Reply:
Yeah, `copy-from` will overwrite the style with whatever you give it. You can put anything in the `style` field, really. And the `display` function can decide how to interpret it if it detects that the widget has a `phrases` array.

Reply:
Hangon, I think I know why I got confused, and I hope you'll indulge me a lengthy explanation because I just realized it's an important oversight in the widget implementation.

There's something in the widget class that I stuck there as part of my original idea for phrase widgets, but which is basically dead code. It's this `widget::text` function, invoked by `widget::value_string` for "text" style widgets.

https://github.com/CleverRaven/Cataclysm-DDA/blob/62a046c52af118778006efd565aedb2d019cc9cf/src/widget.cpp#L765-L768

My thinking back then was to define a field something like `"strings": [ "Peckish", "Hungry", "Ravenous" ]`, and these strings would be mapped to 0, 1, 2 - as returned by some (still mythical) function that yielded a simple numerical progression. It's even defined in the header, and the first thing read in at the top of `widget::load`:

https://github.com/CleverRaven/Cataclysm-DDA/blob/62a046c52af118778006efd565aedb2d019cc9cf/src/widget.h#L159-L160

https://github.com/CleverRaven/Cataclysm-DDA/blob/62a046c52af118778006efd565aedb2d019cc9cf/src/widget.cpp#L223-L225

But `widget::text` is never invoked, because `widget::show` circumvents `color_value_string` for any widget that matches `uses_text_function` - in other words *all* the `*_text` widgets I've defined since then, because it was just easier that way.

https://github.com/CleverRaven/Cataclysm-DDA/blob/62a046c52af118778006efd565aedb2d019cc9cf/src/widget.cpp#L409-L411

Now, I like the `"phrases"` data structure a lot better than a naked list of `"strings"`, but this may be the opportunity to hook up that `widget::text` function for the dynamic rendering it was always destined for. All the dirty work is already done by your `widget_phrase` class. Would it be feasible to hook into place, you think?

Reply:
> Would it be feasible to hook into place, you think?

Absolutely. I can connect the dots in a follow-up PR. It will probably pave the way for #54585.

> [...] this may be the opportunity to hook up that widget::text function for the dynamic rendering [...]

That's a great idea, and along the lines of what I'd like to do as well. It could work like a convenience function to wrap around the phrase functions.

My thought for `widget_phrase` was a simple container for fields, which are accessed via static methods:

```
widget_phrase::get_<field>_for_<condition>( <conditions>, <widget to search> );
```
```c++
widget_phrase::get_text_for_id( const std::string &phrase_id, const widget_id &wgt );
```

The widget functions could wrap these for convenience. The main reason I'm using static accessors for `widget_phrase` is because I wanted to avoid passing the widget itself to the `display` functions.